compiler/src/dotty/tools/dotc/semanticdb/TypeOps.scala

```diff
@@ -0,0 +1,344 @@
+      def enterParamRef(tpe: Type): Unit =
+        tpe match {
+          case lam: LambdaType =>
+            // Find the "actual" binder type for nested LambdaType
+            // For example, `def foo(x: T)(y: T): T` and for `<y>.owner.info` would be like
+            // `MethodType(...<x>, resType = MethodType(...<y>, resType = <T>))`.
+            // (Let's say the outer `MethodType` "outer", and `MethodType` who is
+            // `resType` of outer "inner")
+            //
+            // We try to find the "actual" binder of <y>: `inner`,
+            // and register them to the symbol table with `(<y>, inner) -> <y>`
+            // instead of `("y", outer) -> <y>`
+            if lam.paramNames.contains(sym.name) then
+              paramRefSymtab((lam, sym.name)) = sym
+            else
+              enterParamRef(lam.resType)
+
+          // for class constructor
+          // class C[T] { ... }
+          case cls: ClassInfo if sym.info.isInstanceOf[LambdaType] =>
+            val lam = sym.info.asInstanceOf[LambdaType]
+            cls.cls.typeParams.foreach { param =>
+              paramRefSymtab((lam, param.name)) = param
+            }
+
+          // type X[T] = ...
+          case tb: TypeBounds =>
+            enterParamRef(tb.lo)
+            enterParamRef(tb.hi)
+
+          case _ => ()
+        }
+
+      def enterRefined(tpe: Type): Unit =
+        tpe match {
+          case refined: RefinedType =>
+            val key = (refined, sym.name)
+            refinementSymtab(key) = sym
+
+          case rec: RecType =>
+            enterRefined(rec.parent)
+
+          case cls: ClassInfo
+            if (cls.cls.name == tpnme.REFINE_CLASS) =>
+            sym.owner.owner.info match {
+              // def foo(x: Parent { refinement }) = ...
+              case refined: RefinedType =>
+                enterRefined(refined)
+
+              // type x = Person { refinement }
+              case tb: TypeBounds =>
+                // tb = TypeBounds(
+                //   lo = RefinedType(...)
+                //   hi = RefinedType(...)
+                // )
+                enterRefined(tb.lo)
+                enterRefined(tb.hi)
+
+              // def s(x: Int): { refinement } = ...
+              case expr: ExprType =>
+                enterRefined(expr.resType)
+              case m: MethodType =>
+                enterRefined(m.resType)
+              case _ => ()
+            }
+          case _ => ()
+        }
+      enterParamRef(sym.owner.info)
+      enterRefined(sym.owner.info)
+
+      def loop(tpe: Type): s.Signature = tpe match {
+        case mp: MethodOrPoly =>
+          def flatten(
+            t: Type,
+            paramss: List[List[Symbol]],
+            tparams: List[Symbol]
+          ): (Type, List[List[Symbol]], List[Symbol]) = t match {
+            case mt: MethodType =>
+              val syms = mt.paramNames.flatMap { paramName =>
```

Comment:
I found some bugs by making crush if we couldn't find the symbol from the table
- https://github.com/lampepfl/dotty/pull/12885/commits/907a386bdf8331ea9312652d4759a391dee8df06
- https://github.com/lampepfl/dotty/pull/12885/commits/08d2382f343d560b74127aa84b9d510b0fe771c6

However, my concern is about there might be some potential bugs there and I'm afraid of crashing the compiler or the process of building SemanticDB just because of the bug here.
(And we still have a problem with looking up the symbol for export clause https://github.com/lampepfl/dotty/pull/12885#discussion_r665115632 )

What do you think about just showing a warning message instead of crashing when we failed to lookup symbol?

Reply:
a warning would be better I think - where the message should make it clear it is some internal implementation failure during SemanticDB analysis

Reply:
Emit a warning if symbol lookup failed https://github.com/lampepfl/dotty/pull/12885/commits/09e62eb35a607cce3c35ee3cd5442eb4fac46c5e

The warning looks like
`Internal error in extracting SemanticDB while compiling /Users/tanishiking/src/github.com/tanishiking/dotty/tests/semanticdb/expect/exports-package.scala: Ignoring T of type HKTypeLambda(List(T), List(TypeBounds(TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Nothing),TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Any))), AppliedType(TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class exports)),object example),trait Encoder),List(TypeParamRef(T))))`

Reply:
looks great, thank you